RFT(est): Accurate coverage with inlining on

yuyichao opened the pull request:
With https://github.com/JuliaLang/julia/pull/18163 we are pretty close to coverage with inlining on, the main issue is that type inference delete some of those info during inlining. This PR makes sure those info are preserved. This will cause slightly more memory usage in the IR but hopefully won't change performance (we should ignore metadata during inlining IIRC). This increases the sysimg rodata size from 24MB to 28MB so I guess I need to check if there's any easy case I can optimize without loosing debug info.

For my simple test case, this PR with inlining on produce the same result as that with inlining off and is more accurate than current master with inlining on or off.

This PR (this is as correct as it's ever going to be in this case)

``` jl
        - macro m(x)
        1     quote
        1         c = $(esc(x))
        1         return c + 100
        -     end
        - end
        - 
        1 @inline k(a) = a
        - 
        - @inline function f(a)
        1     c = k(1)
        1     b = a + c
        1     return @m b
        - end
        - 
        - function g(a)
        1     f(a)
        - end
        - 
        - # @code_warntype k(1)
        - # @code_warntype f(1)
        - # @code_warntype g(1)
        - 
        - g(1)
```

Current master inlining off (it doesn't miss any line, but over count some, this is likely fixed by #18163)

``` jl
        - macro m(x)
        1     quote
        2         c = $(esc(x))
        2         return c + 100
        -     end
        - end
        - 
        1 @inline k(a) = a
        - 
        - @inline function f(a)
        1     c = k(1)
        2     b = a + c
        1     return @m b
        - end
        - 
        - function g(a)
        1     f(a)
        - end
        - 
        - # @code_warntype k(1)
        - # @code_warntype f(1)
        - # @code_warntype g(1)
        - 
        - g(1)
```

Current master with inlining on (it tried ;-p )

``` jl
        - macro m(x)
        1     quote
        4         c = $(esc(x))
        -         return c + 100
        -     end
        - end
        - 
        - @inline k(a) = a
        - 
        - @inline function f(a)
        -     c = k(1)
        2     b = a + c
        1     return @m b
        - end
        - 
        - function g(a)
        1     f(a)
        - end
        - 
        - # @code_warntype k(1)
        - # @code_warntype f(1)
        - # @code_warntype g(1)
        - 
        - g(1)
```

Fix #17476 

Does nanosoldier work from top comment?

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

yuyichao committed:
Preserve all line number metadata when coverage is on.

More accurate coverage with inlining on.

yuyichao commented:
The `testsystem` label reminds me that this doesn't actually change the test runners to enable inlining with coverage enabled.

kshyatt commented:
See? I'm helping!

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18196) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d964c98c1f9084ba8704c027bbbcae1da97891ee/0cfda58_vs_3f29b37/report.md). cc @jrevels

yuyichao commented:
I can't reproduce some of the performance regression after fixing a bug in https://github.com/JuliaLang/julia/pull/18163, so let's try again.....

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18196#issuecomment-241691365) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d3d9d38a7804521ba1f8d102b16f9932d0d4eedf/40de579_vs_3f29b37/report.md). cc @jrevels

yuyichao commented:
The inlining cost calculation should be insensitive to additional metadata nodes now so hopefully most of the performance regressions are gone....

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18196#issuecomment-241790541) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/26abec77018609c45500c8e97607e05c12a1f12a/a8f1fce_vs_2390055/report.md). cc @jrevels

JeffBezanson commented:
If this increases sysimg size by 16%, maybe it should only be enabled when `--code-coverage` is passed?

KristofferC commented:
Do you think it was the `inline_worthy` tweaking that gave the performance improvements?

yuyichao commented:
> If this increases sysimg size by 16%, maybe it should only be enabled when `--code-coverage` is passed?

Yes, that's what I want to try. There's also a few more cases we can delete some of the expressions earlier (e.g. move handling of `inbounds`/`boundscheck` to inference after inlining)

yuyichao commented:
With the new commits (which are submitted separately as https://github.com/JuliaLang/julia/pull/18260) the sysimg size without coverage on is back down (to slightly smaller than before due to the added optimizations) I'll also make the change in this PR conditional on coverage later so that we don't need to add the metadata and then delete them soon later....

This obviously still need more testing, a few known cases that are still problematic,

1. `jlcall_api == 2`

    They are problematic in many ways, either inlined by codegen, or interpreted/dynamically dispatched, since they usually don't emit code and therefore won't trigger the coverage collection.

    An easy way to solve this problem is to stop generating `jlcall_api == 2` if coverage is on. In general though, we still need a better way to handle coverage of interpreted code. This is also an issue for backtraces if we starts to interprete some functions instead of doing codegen on all of them.

2. Functions inlined into base function

    For example, when passing a callback to base functions like `map`/`boardcast`. This is only an issue for `--code-coverage=user` and we either need a better way to distinguish inlined user code inside base code or we can just remove `--code-coverage=user`

yuyichao committed:
Track the module during inlining

Compute per-statement `in_user_code` to support user code only logging mode
with inlining on.

yuyichao committed:
Do not disable inlining when coverage is enabled.

Fix #17476

yuyichao commented:
All the issues I've noticed are fixed. With inlining on and user code only coverage my test case gives.

```jl
        - #!/usr/bin/julia -f
        - 
        - macro m(x)
        1     quote
        1         c = $(esc(x))
        1         return c + 100
        -     end
        - end
        - 
        1 @inline k(a) = 2
        - 
        - @inline function f(a)
        1     c = k(1)
        1     b = a + c
        1     return @m b
        - end
        - 
        - @inline function g(a)
        1     f(a)
        - end
        - 
        - # @code_warntype k(1)
        - # @code_warntype f(1)
        - # @code_warntype g(1)
        - 
        - # g(1)
        1 m() = foreach(g, 1)
        - # @code_warntype m()
        - m()
        - 
```

yuyichao commented:
Local testing on my package shows that the coverage is as accurate as (if not more accurate than) what it was with inlining on.

The PR should be ready now and I'd like to merge this some time next week. I encourage people with packages that has poor performance without inlining to check if this helps with the performance problem and if this still give accurate coverage.

Note that this will still give worse performance than no-coverage since,

1. `jlcall_api == 2` is disabled with coverage on. (Hopefully the inliner and LLVM can do a decent job optimizing it)
2. The code that updates the counter also take time to execute.
3. The volatile load/store used by the counter update will disable vectorization.

yuyichao committed:
Fix debug info tracking with non-linear codegen.

tkelman commented on base/inference.jl:
```diff
@@ -1933,8 +1934,11 @@ function finish(me::InferenceState)
 
-    if (isa(me.bestguess,Const) && me.bestguess.val !== nothing) ||
-        (isType(me.bestguess) && !has_typevars(me.bestguess.parameters[1],true))
+    # Do no emit `jlcall_api == 2` if coverage is enabled so that we don't
```
Do not emit

yuyichao commented:
Will merge soon.

The request for test is still valid after the merge. I guess there isn't many 0.6 only packages (or packages that is only tested on 0.6) yet so this won't have a big impact on the coverage report immediately. Hopefully being available in nightly can encourage people to test locally.

yuyichao commented:
FixedPointNumbers test runtime:

Normal (master/this pr): 30s
Coverage (master + noinline): 512s
Coverage (this pr + inlinine): 33s
